### PR TITLE
Pabowers/monitoring python library

### PR DESCRIPTION
**Pull request by pabowers**
I have added the command to add monitoring to a blade. I have confirmed that it calls the correct endpoint in the hanarp, but have not validated against a real blade.

### CODE REVIEW

File: azext_hanaonazure/_params.py [hunk 1]
```diff
@@ -11,4 +11,11 @@
 def load_arguments(self, _):
     with self.argument_context('hanainstance') as c:
         c.argument('resource_group_name', arg_type=resource_group_name_type)
-        c.argument('instance_name', options_list=['--instance-name', '-n'], help="The name of the SAP HANA instance", id_part='name')
\ No newline at end of file
+        c.argument('instance_name', options_list=['--instance-name', '-n'], help="The name of the SAP HANA instance", id_part='name')
+        c.argument('hana_vnet', options_list=['--hana-vnet', '-hv'], help="ARM ID of an Azure Vnet with access to the HANA instance.")
```
Comment (lagalbra) [inline azext_hanaonazure/_params.py]: Do you need `id_part='` for all of these?

**Comment by lagalbra** [comment]
I've never added a CLI command in this repo (and I'm not sure who has) - Donald is my best guess. Whomever has should review.

File: azext_hanaonazure/_params.py [hunk 2]
```diff
@@ -11,4 +11,11 @@
...
+        c.argument('hana_hostname', options_list=['--hana-hostname'], help="Hostname of the HANA Instance blade.")
+        c.argument('hana_instance_num', options_list=['--hana-instance-num'], help="A number between 00 and 99, stored as a string to maintain leading zero.")
+        c.argument('db_container', options_list=['--db-container'], help="Either single or multiple depending on the use of MDC(Multiple Database Containers). Possible values include: 'single','multiple'")
+        c.argument('hana_database', options_list=['--hana-database', '-hdb'], help="Name of the database itself.  It only needs to be specified if using MDC")
+        c.argument('hana_db_username', options_list=['--hana-db-username'], help="Username for the HANA database to login to for monitoring")
+        c.argument('hana_db_password', options_list=['--hana-db-password', '-hdbpw'], help="Password for the HANA database to login for monitoring")
```
Comment (PakDLiu) [inline azext_hanaonazure/_params.py]: We might not want to let the user provide the password like this. It will show a password on their shell history.
We should have the user enter this in much like when you ssh, where the thing you type is not visible.
Or specify a file with the password in it.

Check to see if other commands to this. I am not sure how to do this off the top of my head.

**Review by PakDLiu — CHANGES_REQUESTED** [review]

File: README.md [hunk 3]
```diff
@@ -45,9 +45,16 @@ To delete all key-value pairs in the Tags field of a specific HANA instance:
 ```
 az hanainstance update --resource-group $RESOURCE_GROUP --instance-name $HANA_INSTANCE_NAME --set tags={}
 ```
+
+To enable monitoring for a specific HANA instance:
+
+```
+az hanainstance monitor --resource-group $RESOURCE_GROUP --instance-name $HANA_INSTANCE_NAME --hana-vnet $HANA_VNET  --hana-hostname $HANA_HOSTNAME --hana-instance-num $HANA_INSTANCENUM --hana-db-username $HANA_DB_USERNAME --hana-db-password $HANA_DB_PW
```
Comment (tniek) [inline README.md]: I may have missed this, but was the method initially not "enable monitoring" or "toggle monitoring"?

Comment (tniek) [reply README.md]: Also, what exactly is the instance name and why is it the only HANA-specific parameter that is not prefixed by "HANA"?

File: azext_hanaonazure/_params.py [hunk 4]
```diff
@@ -11,4 +11,11 @@
 def load_arguments(self, _):
     with self.argument_context('hanainstance') as c:
         c.argument('resource_group_name', arg_type=resource_group_name_type)
-        c.argument('instance_name', options_list=['--instance-name', '-n'], help="The name of the SAP HANA instance", id_part='name')
\ No newline at end of file
+        c.argument('instance_name', options_list=['--instance-name', '-n'], help="The name of the SAP HANA instance", id_part='name')
```
Comment (tniek) [inline azext_hanaonazure/_params.py]: See above; it is not clear what is meant by "name of the SAP HANA instance": The SID? The resource name? An arbitrary name the user can assign to a SAP HANA instance?

File: azext_hanaonazure/_params.py [hunk 5]
```diff
@@ -11,4 +11,11 @@
...
+        c.argument('hana_vnet', options_list=['--hana-vnet', '-hv'], help="ARM ID of an Azure Vnet with access to the HANA instance.")
+        c.argument('hana_hostname', options_list=['--hana-hostname'], help="Hostname of the HANA Instance blade.")
+        c.argument('hana_instance_num', options_list=['--hana-instance-num'], help="A number between 00 and 99, stored as a string to maintain leading zero.")
+        c.argument('db_container', options_list=['--db-container'], help="Either single or multiple depending on the use of MDC(Multiple Database Containers). Possible values include: 'single','multiple'")
```
Comment (tniek) [inline azext_hanaonazure/_params.py]: I suppose this refers to the (boolean) parameter `HanaMdc`; however, `db_container`, is misleading as it suggests that a specific DB container is expected (rather than a boolean whether MDC is used or not).

File: azext_hanaonazure/commands.py [hunk 6]
```diff
@@ -17,5 +17,6 @@ def load_command_table(self, _):
         g.custom_command('list', 'list_hanainstance')
         g.custom_command('show', 'show_hanainstance', exception_handler=empty_on_404)
         g.custom_command('restart', 'restart_hanainstance')
+        g.custom_command('monitor', 'enable_monitoring_hanainstance')
```
Comment (tniek) [inline azext_hanaonazure/commands.py]: Again, "monitor" refers to the action of monitoring itself (rather than enabling the monitor).

File: azext_hanaonazure/custom.py [hunk 7]
```diff
@@ -28,3 +28,17 @@ def restart_hanainstance(client, resource_group_name, instance_name):
...
     return client.update(resource_group_name, instance_name, kwargs['parameters'].tags)
+
+def enable_monitoring_hanainstance(client, resource_group_name, instance_name, hana_vnet, hana_hostname, hana_instance_num, hana_db_username, hana_db_password, db_container="single", hana_database=None):
+    monitoring_details = {
+        "hanaInstanceNum": hana_instance_num,
+        "hanaVnet": hana_vnet,
+        "hanaHostname": hana_hostname,
+        "hanaDbUsername": hana_db_username,
+        "hanaDbPassword": hana_db_password,
+        "dbContainer": db_container
```
Comment (tniek) [inline azext_hanaonazure/custom.py]: Again, not a fan of `dbContainer` vs. a boolean `HanaMdc`:
`dbContainer` seems to refer more to the container (aka the tenant DB itself).

File: azext_hanaonazure/modules_sdk/models/hana_management_client_enums.py [hunk 8]
```diff
@@ -48,3 +48,9 @@ class HanaInstancePowerStateEnum(str, Enum):
     stopped = "stopped"
     restarting = "restarting"
     unknown = "unknown"
+
+
+class HanaDatabaseContainersEnum(str, Enum):
```
Comment (tniek) [inline azext_hanaonazure/modules_sdk/models/hana_management_client_enums.py:53]: See above; this should be a boolean whether the specific HDB instance is MDC or not. The whole purpose is to provide downward support for (legacy) single-tenant DBs.

File: azext_hanaonazure/modules_sdk/models/monitoring_details.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        'hana_db_username': {'key': 'hanaDbUsername', 'type': 'str'},
+        'hana_db_password': {'key': 'hanaDbPassword', 'type': 'str'},
+    }
+
+    def __init__(self, **kwargs):
+        super(MonitoringDetails, self).__init__(**kwargs)
+        self.hana_vnet = kwargs.get('hana_vnet', None)
+        self.hana_hostname = kwargs.get('hana_hostname', None)
+        self.hana_instance_num = kwargs.get('hana_instance_num', None)
+        self.db_container = kwargs.get('db_container', "single")
```
Comment (tniek) [inline azext_hanaonazure/modules_sdk/models/monitoring_details.py:57]: The default behavior should be to support MDCs (since this is the standard as of HANA1.0 SPS12 and HANA2.0); so default should be `multiple` (or `true` if we go with a boolean `HanaMdc`).

**Review by tniek — COMMENTED** [review]
Most of my comments refer to the use of a string/enum to determine whether a specific HDB instance is MDC (default behavior) or single-tenant (legacy). I find the naming `db_container` (instead of `HanaMdc`) confusing, since it refers more to an actual container/tenant than the mode of DB itself.

File: README.md (hunk 3 above)
Comment (pabowers) [reply README.md]: instance name is the name of the `hanainstance` since it is used in the other methods as `instance-name` I kept it that way for consistency. It may make sense to remove the prefix from the other arguments so that the user does not have to type in such long names

File: azext_hanaonazure/modules_sdk/models/hana_management_client_enums.py (hunk 8 above)
Comment (pabowers) [reply azext_hanaonazure/modules_sdk/models/hana_management_client_enums.py:53]: This is auto-generated code. The reason that it is not a boolean is because the ARM team does not like to allow booleans in their configuration code. I thought that we had decided on this naming over a month ago when I was getting the swagger spec approved.

File: azext_hanaonazure/modules_sdk/models/monitoring_details.py (hunk 9 above)
Comment (pabowers) [reply azext_hanaonazure/modules_sdk/models/monitoring_details.py:57]: auto-generated code

Comment (pabowers) [reply azext_hanaonazure/modules_sdk/models/monitoring_details.py:57]: If you would like this changed, it needs to be changed in the swagger spec and not in this PR

File: azext_hanaonazure/custom.py (hunk 7 above)
Comment (pabowers) [reply azext_hanaonazure/custom.py]: This is the one place where I can actually change this in this PR.

File: azext_hanaonazure/_params.py (hunk 4 above)
Comment (pabowers) [reply azext_hanaonazure/_params.py]: I don't actually make any change to this line. This is there before. The only reason that it marks the line as changed is because I added a newline after it to add the monitoring parameters

File: azext_hanaonazure/commands.py (hunk 6 above)
Comment (pabowers) [reply azext_hanaonazure/commands.py]: I'll change to `enable-monitoring`

File: azext_hanaonazure/_params.py (hunk 1 above)
Comment (pabowers) [reply azext_hanaonazure/_params.py]: I believe that only pieces of the ARM id have that part

File: azext_hanaonazure/_params.py (hunk 2 above)
Comment (pabowers) [reply azext_hanaonazure/_params.py]: This is the same way that it is done for the virtual machine creation with a password https://github.com/Azure/azure-cli/blob/a32eb7156d7e814d38e2374e77fbd4d131bfe7f6/src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_params.py. I agree with your point, but not sure how to adjust while still using the library

File: azext_hanaonazure/_params.py (hunk 5 above)
Comment (pabowers) [reply azext_hanaonazure/_params.py]: I have renamed it

File: azext_hanaonazure/_params.py (hunk 2 above)
Comment (jovillas) [reply azext_hanaonazure/_params.py]: -p is pretty standard... but I would say that if -p is not present, it the command should expect it from stdin, that way the user can 
cmd < /path/to/password_file

File: README.md [hunk 10]
```diff
@@ -31,7 +31,7 @@ az hanainstance update --resource-group $RESOURCE_GROUP --instance-name $HANA_IN
 To update a key-value pair in the Tags field of a specific HANA instance:
 
 ```
-az hanainstance update --resource-group $RESOURCE_GROUP --instance-name $HANA_INSTANCE_NAME --set tags.key=udpatedValue
+az hanainstance update --resource-group $RESOURCE_GROUP --instance-name $HANA_INSTANCE_NAME --set tags.key=updatedValue
```
Comment (jovillas) [inline README.md]: What changed on this line? Is it related to monitoring?

File: README.md [hunk 11]
```diff
@@ -45,9 +45,16 @@ To delete all key-value pairs in the Tags field of a specific HANA instance:
 ```
 az hanainstance update --resource-group $RESOURCE_GROUP --instance-name $HANA_INSTANCE_NAME --set tags={}
 ```
+
+To enable monitoring for a specific HANA instance:
+
+```
+az hanainstance enable-monitoring --resource-group $RESOURCE_GROUP --instance-name $HANA_INSTANCE_NAME --hana-vnet $HANA_VNET  --hana-hostname $HANA_HOSTNAME --hana-instance-num $HANA_INSTANCENUM --hana-db-username $HANA_DB_USERNAME --hana-db-password $HANA_DB_PW
```
Comment (jovillas) [inline README.md]: There is an extra space between `$HANA_VNET` and `--hana-hostname`

I suggest to split this line to improve readability.

File: azext_hanaonazure/commands.py [hunk 12]
```diff
@@ -9,13 +9,16 @@
 from ._client_factory import cf_hanainstance_groups
 from azure.cli.core.commands import CliCommandType
 
+
```
Comment (jovillas) [inline azext_hanaonazure/commands.py]: Remove extra line.

File: azext_hanaonazure/commands.py [hunk 13]
```diff
@@ -9,13 +9,16 @@
...
     custom_tmpl = 'azext_hanaonazure.custom#{}'
     custom_type = CliCommandType(operations_tmpl=custom_tmpl)
 
     with self.command_group('hanainstance', client_factory=cf_hanainstance_groups) as g:
         g.custom_command('list', 'list_hanainstance')
-        g.custom_command('show', 'show_hanainstance', exception_handler=empty_on_404)
+        g.custom_command('show', 'show_hanainstance',
+                         exception_handler=empty_on_404)
         g.custom_command('restart', 'restart_hanainstance')
+        g.custom_command('enable-monitoring', 'enable_monitoring_hanainstance')
```
Comment (jovillas) [inline azext_hanaonazure/commands.py]: Is there going to be a `disable-monitoring`?

File: azext_hanaonazure/_params.py [hunk 14]
```diff
@@ -11,4 +11,19 @@
 def load_arguments(self, _):
     with self.argument_context('hanainstance') as c:
         c.argument('resource_group_name', arg_type=resource_group_name_type)
-        c.argument('instance_name', options_list=['--instance-name', '-n'], help="The name of the SAP HANA instance", id_part='name')
\ No newline at end of file
+        c.argument('instance_name', options_list=[
+                   '--instance-name', '-n'], help="The name of the SAP HANA instance", id_part='name')
+        c.argument('hana_vnet', options_list=[
```
Comment (jovillas) [inline azext_hanaonazure/_params.py:17]: What happens if the user adds this parameters for other commands?

File: azext_hanaonazure/commands.py (hunk 13 above)
Comment (jovillas) [inline azext_hanaonazure/commands.py]: It seems to me that `monitoring` is a sub-command which have `enable`, `disable` actions and all the monitoring specific parameters to only be valid to this sub-command.

File: azext_hanaonazure/custom.py [hunk 15]
```diff
@@ -28,3 +28,17 @@ def restart_hanainstance(client, resource_group_name, instance_name):
...
+def enable_monitoring_hanainstance(client, resource_group_name, instance_name, hana_vnet, hana_hostname, hana_instance_num, hana_db_username, hana_db_password, hana_mdc=True, hana_database=""):
+    monitoring_details = {
+        "hanaInstanceNum": hana_instance_num,
+        "hanaVnet": hana_vnet,
+        "hanaHostname": hana_hostname,
+        "hanaDbUsername": hana_db_username,
+        "hanaDbPassword": hana_db_password,
+        "dbContainer": "multiple" if hana_mdc else "single"
+    }
+    if hana_database is not None:
```
Comment (jovillas) [inline azext_hanaonazure/custom.py:42]: Is it valid to `enable-monitoring` without a `hana_database`? What would this be monitoring in that case?

Actually, what does the hanaRP API will reply in this scenario? I expect some "invalid parameters" error.

Comment (jovillas) [reply azext_hanaonazure/custom.py:42]: Nevermind, I just saw the MonitoringDetails docs

**Review by PakDLiu — APPROVED** [review]

File: azext_hanaonazure/modules_sdk/models/tags.py [hunk 16]
```diff
@@ -20,9 +20,9 @@ class Tags(Model):
     """
 
     _attribute_map = {
-    'tags': {'key': 'tags', 'type': '{str}'},
+        'tags': {'key': 'tags', 'type': '{str}'},
```
Comment (jovillas) [inline azext_hanaonazure/modules_sdk/models/tags.py:23]: Why was this file changed?

File: azext_hanaonazure/modules_sdk/models/tags_py3.py [hunk 17]
```diff
@@ -25,4 +25,4 @@ class Tags(Model):
 
     def __init__(self, *, tags=None, **kwargs) -> None:
         super(Tags, self).__init__(**kwargs)
-        self.tags = tags
\ No newline at end of file
+        self.tags = tags
```
Comment (jovillas) [inline azext_hanaonazure/modules_sdk/models/tags_py3.py:28]: Same, why was this file touched?

**Review by jovillas — CHANGES_REQUESTED** [review]

File: README.md (hunk 10 above)
Comment (pabowers) [reply README.md]: not related to monitoring, I found a typo and thought I would fix. I can fix in a different PR, but thought it would be ok to make the change here since it's so small. I just changed "udpatedValue" to "updatedValue"

File: azext_hanaonazure/commands.py (hunk 13 above)
Comment (pabowers) [reply azext_hanaonazure/commands.py]: I believe that that is the eventual goal, but we don't have it now

File: azext_hanaonazure/modules_sdk/models/tags_py3.py (hunk 17 above)
Comment (pabowers) [reply azext_hanaonazure/modules_sdk/models/tags_py3.py:28]: this was auto-generated.

File: azext_hanaonazure/_params.py (hunk 14 above)
Comment (pabowers) [reply azext_hanaonazure/_params.py:17]: nothing happens in that case. It only shows up for this command

File: azext_hanaonazure/modules_sdk/models/tags.py (hunk 16 above)
Comment (pabowers) [reply azext_hanaonazure/modules_sdk/models/tags.py:23]: not sure, it was auto-generated

File: azext_hanaonazure/_params.py (hunk 14 above)
Comment (pabowers) [reply azext_hanaonazure/_params.py:17]: I did add an argument context, so now they will definitely only show up under `hanainstance enable-monitoring`

File: README.md (hunk 10 above)
Comment (jovillas) [reply README.md]: Yeah, I prefer this kind of things to be on a different PR to prevent it being reverted if this commit ever needs to be reverted.

File: azext_hanaonazure/modules_sdk/models/tags.py (hunk 16 above)
Comment (jovillas) [reply azext_hanaonazure/modules_sdk/models/tags.py:23]: Shame on autogenerating tools.

File: README.md (hunk 10 above)
Comment (pabowers) [reply README.md]: ok, will remove from here and send out another PR to fix that

File: azext_hanaonazure/_params.py (hunk 14 above)
Comment (jovillas) [reply azext_hanaonazure/_params.py:17]: I still think it will be cleaner to have a `monitor` sub-command. Then have the arguments scoped to that sub-command. See az network nsg/dns/...

File: azext_hanaonazure/commands.py (hunk 13 above)
Comment (jovillas) [reply azext_hanaonazure/commands.py]: See the comment on the  `monitor` sub-command.

Comment (jovillas) [reply azext_hanaonazure/commands.py]: Thinking more about the monitoring feature and talking whit few folks around. I strongly believe the command structure should separate the monitoring pieces from the rest of the `hanainstance` ones.

This is the proposal:
`az hanainstance monitor enable` to enable monitoring on the hanainstance
`az hanainstance monitor disable` once the disable part is done, have this command here.
have all the parameters that are only relevant to the monitoring showing only for this sub-command.

Comment (pabowers) [reply azext_hanaonazure/commands.py]: done

**Review by jovillas — APPROVED** [review]
lgtm
